.github/workflows/binary.yml

```diff
@@ -0,0 +1,290 @@
+          if [ -n "${ENABLED_CONFIGS}" ]; then
+            echo "C graph-lang found for configs:" $(ls config-list)
+          else
+            echo "No C graph-lang found, nothing to do"
+          fi
+      - name: Initialise output artifact
+        if: ${{ steps.prepare.outputs.enabled_configs }}
+        uses: actions/upload-artifact@v3
+        with:
+          name: graph-refine-targets
+          path: |
+            decompile-manifest.yaml
+            verification-manifest.xml
+            configs
+
+  decompilation:
+    name: Decompile
+    needs: artifacts
+    runs-on: ubuntu-latest
+    # `if` applies to the whole matrix, not to individual jobs within the matrix.
+    if: ${{ needs.artifacts.outputs.enabled_configs }}
+    strategy:
+      fail-fast: false
+      matrix:
+        arch: [ARM, RISCV64]
+        features: ["", MCS]
+        optimise: ["-O1", "-O2"]
+    steps:
+      # It would be nice if there was a way to prevent the job from starting.
+      - name: Check enabled
+        id: enabled
+        shell: bash
+        env:
+          ARCH: ${{ matrix.arch }}
+          FEATURES: ${{ matrix.features }}
+          OPTIMISE: ${{ matrix.optimise }}
+          ENABLED: ${{ needs.artifacts.outputs.enabled_configs }}
+        run: |
+          # Check whether this configuration is enabled
+          CONFIG="${ARCH}${FEATURES:+-${FEATURES}}"
+          if grep -qx "${CONFIG}" <<< "${ENABLED}"; then
+            echo "C graph-lang found for ${CONFIG}, proceeding with decompilation"
+            echo "::set-output name=config::${CONFIG}"
+            echo "::set-output name=target::${CONFIG}${OPTIMISE}"
+          else
+            echo "No C graph-lang found for ${CONFIG}, skipping decompilation"
+          fi
+      - name: Checkout l4v
+        uses: actions/checkout@v3
+        # We use the ref on which this workflow was triggered,
+        # not the one that caused the trigger.
+        with:
+          path: l4v
+      - name: Checkout Isabelle
+        uses: actions/checkout@v3
+        with:
+          repository: seL4/isabelle
+          ref: ${{ needs.artifacts.outputs.isabelle_commit }}
+          path: isabelle
+      - name: Checkout seL4
+        uses: actions/checkout@v3
+        with:
+          repository: seL4/seL4
+          ref: ${{ needs.artifacts.outputs.sel4_commit }}
+          path: seL4
+      - name: Checkout graph-refine
+        uses: actions/checkout@v3
+        with:
+          repository: seL4/graph-refine
+          ref: ${{ needs.artifacts.outputs.graph_refine_commit }}
+          path: graph-refine
+      - name: Install Nix
+        if: steps.enabled.outputs.config
+        uses: cachix/install-nix-action@v16
+        with:
+          nix_path: nixpkgs=channel:nixos-unstable
+      - name: Install Cachix
+        if: steps.enabled.outputs.config
+        uses: cachix/cachix-action@v10
+        with:
+          name: sel4-bv
+          authToken: ${{ secrets.BV_CACHIX_AUTH_TOKEN }}
```

Comment:
What does this part do?

Reply:
This was always going to be the most controversial bit. I should have added some notes in the commit message, and also in some comments in the workflow. Here's some explanation for a start...

`install-nix-action` installs the [Nix package manager](https://nixos.org/manual/nix/stable/), which provides various `nix-*` commands, and sets it up to use the [nixpkgs collection of packages](https://nixos.org/manual/nixpkgs/stable/). The only command I'm using in this workflow is `nix-shell`, which creates a sub-shell with the environment specified by [`shell.nix`](https://github.com/seL4/graph-refine/blob/ci-riscv64/shell.nix) in the working directory. The environment created by that Nix expression provides the Python modules, cross compilers and other tools the workflow script needs. `nix-shell` will download the packages it needs to set up that environment, building if necessary.

The motivation is to have a declarative specification of the required environment, and a tool to quickly get a script running in that environment. A Dockerfile would serve the same purpose, but Nix is more declarative and makes it easier to ensure build reproducibility. That makes caching work better and speeds up development iteration. Nix also has a very nice way to turn any set of packages [into a Docker image](https://github.com/seL4/graph-refine/blob/ci-riscv64/decompiler/default.nix#L81-L85). I used that to make Docker images for the decompiler and graph-refine. I would have liked to make a Docker image for everything needed here, but ran out of time.

`cachix-action` is optional. It adds a cache for any packages that `nix-shell` needs to build, so it doesn't need to build the same package more than once. The cache is provided by cachix.org, currently running on a 5GB free account, which is plenty big enough for these workflows.

Reply:
I added some comments in [a fixup commit](https://github.com/seL4/l4v/pull/443/commits/a432632b2bc558e62e16e545b9082b7acbdaca14).

Reply:
Ok, thanks for those. You're right, this is a bit controversial, because it introduces yet another way of specifying dependencies. I think it's fine for development at the moment and getting things working, but longer term we should unify these.

Good to merge for now, though.

Reply:
I'll leave it open for a day or two in case it needs further tweaks. I'll ping you if there's anything significant.